src/App.js

```diff
@@ -22,25 +22,43 @@ class App extends React.Component {
       description: this.state.newTodoDescription,
       completed: false
     }
-    console.log("createTodo", newTodo)
-
     // TODO: add the new Todo to the list in state
     // make sure not to mutate the existing array!
+// create this part in react
+    this.setState(previous  => ({
+        todos: [...previous.todos, newTodo]
+      })
+    )
+
   }
 
   deleteTodo = id => {
-    console.log("deleteTodo", id)
+    // console.log("deleteTodo", id)
 
     // TODO: remove ONE todo from state using the id
     // make sure not to mutate the existing array!
+    this.setState((prev) => ({
+      todos: prev.todos.filter(todo => todo.id !== id)
+    }))
+
   }
 
   updateTodo = (id, completed) => {
-    console.log("updateTodo", id, completed)
 
     // TODO: update ONE todo in state using the id
     // make sure not to mutate the existing array!
     // also, make sure not to mutate any objects within the array!
+    // look through the array of todos match it to id, and change completed to true
+    this.setState(prev => ({todos: prev.todos.map((todo) => {
+      if (todo.id === id){
+        todo.completed = completed
+        return todo
+      }
+      return todo
+    }
+    )}))
```

Comment:
Be careful when you're working with objects in state that you don't modify the objects directly; technically this still counts as 'mutating state' - on line 54 when you have `todo.completed = complete` this is mutating state directly. Instead, you should return a new todo object (which you can do with the spread operator) and then change its completed status:

However, in this case you *are* mutating state inside of the map function on line 49. Instead, you should create a new object inside the map function for the todo that you're updating, like this:

```js
    prev.todos.map( (todo) => {
      if (todo.id === id) {
        return { ...todo, completed: completed } // spread makes a copy of the object, then we're updating the completed status of the new object
      }
      return todo
    })
```